src/main/java/net/logstash/logback/util/ReusableByteBuffer.java

```diff
@@ -0,0 +1,240 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.util;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * A speedy alternative to {@link java.io.ByteArrayOutputStream}.
+ *
+ * <p>Unlike {@link java.io.ByteArrayOutputStream}, this implementation is backed by an {@link ArrayList}
+ * of {@code byte[]} instead of 1 constantly resizing {@code byte[]} array. It does not copy buffers when
+ * it gets expanded.
+ *
+ * <p>The initial buffer is only created when the stream is first written.
+ * There is also no copying of the internal buffer if its contents is extracted with the
+ * {@link #writeTo(OutputStream)} method.
+ *
+ * <p>The {@link #reset()} method clears the content and resets the buffer to its initial state.
+ * Buffers are disposed except the initial buffer which is reused by subsequent usage.
+ *
+ * @author brenuart
+ *
+ */
+public class ReusableByteBuffer extends OutputStream {
+
+    /**
+     * The default size of the initial buffer
+     */
+    public static final int INITIAL_SIZE = 1024;
```

Comment:
It could... it is currently only used by `ReusableByteBuffers` - so it can be made _package private_ for instance. However, I wonder if it would interesting for others to know the default size ?

Reply:
Let's reduce to package private for now.  We can always increase later if there is a solid use case for it.

Reply:
What about `1024` bytes for the default capacity? Is it large enough? 
It is important to have a good default value here applicable to most use cases. May be we should also add a word about it in the documentation.. ?

Reply:
> What about 1024 bytes for the default capacity? Is it large enough?

Hmmm.  Good question.  I just took a quick glance over our logs, and the majority seem to fit within 1024.

Another question... Let's say this capacity is too low for one particular customer's usage... When they upgrade to this new version of logstash-logback-encoder, will their performance be _worse/comparable/better_ than the old behavior?

> It is important to have a good default value here applicable to most use cases. May be we should also add a word about it in the documentation.. ?

Yeah, probably worth a section in the readme.

> Are you aware of any strategy that would grow/shrink the buffer size dynamically ?

Not without increasing the complexity of this feature tremendously.  Let's use the simple solution first.

Reply:
> Let's say this capacity is too low for one particular customer's usage... When they upgrade to this new version of logstash-logback-encoder, will their performance be worse/comparable/better than the old behavior?

Performance will be similar to old behaviour but not worst. The old implementation relied on Jackson Recycler and related buffers. If you remember our previous discussion on the subject I highlighted the fact that we actually had close to no reuse of the allocated buffers at all. When doing my experiments, I compared the performance of the new buffering mechanism with _reuse disabled_ (buffers were never returned to the pool so a new one had to be created every time) against 6.6. Results were:
- Speed: -5% (slower than 6.6)
- GC count: -2%
- GC time: -1.5%
- Allocations: -1.5%

This is the worst case scenario. As you can see, performances are _similar_ to 6.6...

> Hmmm. Good question. I just took a quick glance over our logs, and the majority seem to fit within 1024.
`1024` was the default `minBufferSize` of `CompositeJsonEncoder`... I took the same default.

Reply:
Ok.  I'm fine with 1024, mainly just based on the logs that I looked at quickly.  I'd be ok doubling it too, but wouldn't want to go too much higher.  Your call.

Reply:
> Yeah, probably worth a section in the readme.
I added some words about it under "Composite Encoders/Layouts". Please review after I commit and don't hesitate to rephrase my English...